syntax = "proto3";
package minecpp.engine;

import "game.proto";
import "player.proto";

service PlayerService {
    rpc AcceptPlayer (AcceptPlayerRequest) returns (AcceptPlayerResponse);
    rpc SetPlayerPosition (SetPlayerPositionRequest) returns (EmptyResponse);
    rpc SetPlayerRotation (SetPlayerRotationRequest) returns (EmptyResponse);
    rpc ChatMessage (ChatMessageRequest) returns (EmptyResponse);
    rpc ListPlayers (EmptyRequest) returns (ListPlayersResponse);
    rpc ListPlayerEntities (EmptyRequest) returns (ListPlayerEntitiesResponse);
    rpc RemovePlayer(RemovePlayerRequest) returns (EmptyResponse);
}

message AcceptPlayerRequest {
    string name = 1;
    bytes uuid = 2;
}

message AcceptPlayerResponse {
    PlayerAcceptState state = 1;
    uint32 node_id = 2;
    minecpp.game.GamePlay game_info = 3;
    minecpp.player.Data player_data = 4;

    enum PlayerAcceptState {
        DENIED = 0;
        ACCEPTED = 1;
        MOVED = 2;
    }
}

message SetPlayerPositionRequest {
    bytes uuid = 1;
    double x = 2;
    double y = 3;
    double z = 4;
}

message SetPlayerRotationRequest {
    bytes uuid = 1;
    float yaw = 2;
    float pitch = 3;
}

message ChatMessageRequest {
    bytes uuid = 1;
    string message = 2;
}

message PlayerData {
    bytes uuid = 1;
    string name = 2;
    minecpp.game.GameMode game_mode = 3;
}

message ListPlayersResponse {
    repeated PlayerData list = 1;
}

message PlayerEntityData {
    bytes uuid = 1;
    int32 entity_id = 2;
    double x = 3;
    double y = 4;
    double z = 5;
    float yaw = 6;
    float pitch = 7;
}

message ListPlayerEntitiesResponse {
    repeated PlayerEntityData list = 1;
}

message RemovePlayerRequest {
    bytes uuid = 1;
}

message EmptyRequest {
}

message EmptyResponse {
}
